#### 데이터 정보
* (60000, 28, 28) : 색이 없던것
* (60000, 28, 28, 3) : RGB 컬러 있음
* 2D : 행렬(타이타닉, 바이크)...
* 3D : 시계열 데이터
* 4D : (샘플개수, 높이, 너비, 채널) 채널1 : 흑백, 채널3(RGB)
* 5D : 동영상(샘플개수, 프레임, 높이, 너비, 채널)

### 텐서의 크기 변환

In [1]:
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [2]:
train_images=train_images.reshape((60000,28*28)) # 3D -> 2D
print(train_images.ndim)
print(train_images.shape)

2
(60000, 784)


### 라이브러리 불러오기

In [5]:
import keras
import pandas as pd 
import numpy as np 

In [6]:
## parse_dates : datetime 컬럼을 시간형으로 불러올 수 있음
train = pd.read_csv("./Data/bike_mod_tr.csv", parse_dates=['datetime'])
test = pd.read_csv("./Data/bike_mod_test.csv", parse_dates=['datetime'])

### 데이터 선택

In [5]:
input_col=['hour', 'temp']
labeled_col=['count']

In [6]:
X=train[input_col]
y=train[labeled_col]
X_val=test[input_col]

In [12]:
from sklearn.model_selection import train_test_split

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

In [9]:
# random seed값 0
seed=0
np.random.seed(seed)

### 딥러닝 구조 결정
* 사용하는 특징(feature) : 2개
* 마지막 예측하는 것이 회귀 - 출력층의 노드는 1개
* 마지막 예측하는 것이 분류 - 출력층 노드는 범주의 개수만큼 생성됨.

* Sequential() 함수를 사용하여 각 층별 쉽게 쌓아올릴 수 있음.
* model.add()로 한 층 또는 두 층씩 추가
* 첫 model.add() 입력층과 그 다음 출력층을 넣을 수 있음.
* 각 은닉층이 끝날 때 활성화 함수를 넣어 비선형 문제를 풀 수 있음.

In [10]:
from keras.models import Sequential
from keras.layers import Dense

In [11]:
model=Sequential()
model.add(Dense(30,input_dim=2, activation='relu'))
model.add(Dense(15,activation='relu'))
model.add(Dense(15,activation='relu'))
model.add(Dense(1))

In [12]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 30)                90        
_________________________________________________________________
dense_1 (Dense)              (None, 15)                465       
_________________________________________________________________
dense_2 (Dense)              (None, 15)                240       
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 16        
Total params: 811
Trainable params: 811
Non-trainable params: 0
_________________________________________________________________


### 배치 데이터 10개
* (10,2)-> 2D

### 평가지표
* MSE, RMSE, RMLSE(또는 RMSLE), MAE
* MSE로 평가한다면 loss='mean_squared_error'
* optimizer: 가중치를 업데이트 시키는데 사용하는 알고리즘(cs231n)

In [13]:
model.compile(loss='mean_squared_error', optimizer='rmsprop')
model.fit(X_train, y_train, epochs=20, batch_size=10)

Epoch 1/20
817/817 [==============================] - 1s 830us/step - loss: 31049.9492
Epoch 2/20
817/817 [==============================] - 1s 773us/step - loss: 24041.0859
Epoch 3/20
817/817 [==============================] - 1s 955us/step - loss: 23002.6875
Epoch 4/20
817/817 [==============================] - 1s 811us/step - loss: 21296.7246
Epoch 5/20
817/817 [==============================] - 1s 792us/step - loss: 19877.7871
Epoch 6/20
817/817 [==============================] - 1s 802us/step - loss: 19534.5957
Epoch 7/20
817/817 [==============================] - 1s 794us/step - loss: 19415.1191
Epoch 8/20
817/817 [==============================] - 1s 773us/step - loss: 19324.0293
Epoch 9/20
817/817 [==============================] - 1s 808us/step - loss: 19229.2773
Epoch 10/20
817/817 [==============================] - 1s 775us/step - loss: 19189.2598
Epoch 11/20
817/817 [==============================] - 1s 767us/step - loss: 19065.3223
Epoch 12/20
817/817 [====================

### 평가 확인

In [14]:
model.evaluate(X_test, y_test)

86/86 [==============================] - 0s 719us/step - loss: 18035.2129


18035.212890625

### 실습 4-1
* 변수 추가를 통해 성능향상 시켜보기.

In [14]:
import keras
import pandas as pd 
import numpy as np 

In [15]:
## parse_dates : datetime 컬럼을 시간형으로 불러올 수 있음
train = pd.read_csv("./Data/bike_mod_tr.csv", parse_dates=['datetime'])
test = pd.read_csv("./Data/bike_mod_test.csv", parse_dates=['datetime'])

In [16]:
from keras.models import Sequential
from keras.layers import Dense

In [17]:
train.columns

Index(['datetime', 'season', 'holiday', 'workingday', 'weather', 'temp',
       'atemp', 'humidity', 'windspeed', 'casual', 'registered', 'count',
       'year', 'month', 'day', 'hour', 'minute', 'second', 'dayofweek'],
      dtype='object')

In [23]:
input_col=['hour', 'atemp', 'humidity', 'windspeed', 'season', 'workingday','holiday']
labeled_col=['count']

X=train[input_col]
y=train[labeled_col]
X_val=test[input_col]

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

model=Sequential()
model.add(Dense(100,input_dim=7, activation='relu'))
model.add(Dense(50,activation='relu'))
model.add(Dense(50,activation='relu'))
model.add(Dense(1))

model.compile(loss='mean_squared_error', optimizer='rmsprop')
model.fit(X_train, y_train, epochs=100, batch_size=10)

Epoch 1/100
817/817 [==============================] - 1s 944us/step - loss: 24673.6992
Epoch 2/100
817/817 [==============================] - 1s 869us/step - loss: 20352.4375
Epoch 3/100
817/817 [==============================] - 1s 927us/step - loss: 18700.1777
Epoch 4/100
817/817 [==============================] - 1s 904us/step - loss: 18078.9492
Epoch 5/100
817/817 [==============================] - 1s 912us/step - loss: 17758.7812
Epoch 6/100
817/817 [==============================] - 1s 900us/step - loss: 17385.1777
Epoch 7/100
817/817 [==============================] - 1s 907us/step - loss: 17137.4180
Epoch 8/100
817/817 [==============================] - 1s 862us/step - loss: 16764.5840
Epoch 9/100
817/817 [==============================] - 1s 846us/step - loss: 16282.9512
Epoch 10/100
817/817 [==============================] - 1s 899us/step - loss: 16116.5352
Epoch 11/100
817/817 [==============================] - 1s 991us/step - loss: 15832.2598
Epoch 12/100
817/817 [========

In [24]:
model.evaluate(X_test, y_test)

86/86 [==============================] - 0s 904us/step - loss: 8128.7266


8128.7265625

### 트라이 1. 8128.7265625
* 변수 : 'hour', 'atemp', 'humidity', 'windspeed', 'season', 'workingday','holiday'
* 은닉층 : 100-50-50-1 
* epochs : 100

### 로그변환 하면 점수 많이 오름.